### PR TITLE
Implement database iterator

### PR DESCRIPTION
**Pull request by jjxtra**
I would appreciate code review on this. It passes all tests.

I would love to see this in a nuget package soon.

As a side note, some of the threading tests are failing. I did not investigate that.

### CODE REVIEW

**Comment by jjxtra** [comment]
I would suggest a squash and merge on this to avoid soiling the commit history

File: MaxMind.Db/CachedDictionary.cs [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+        protected virtual void OnCachedValueRemoved(TKey key, TValue value) { }
+
+        #endregion Protected methods
+
+        #region Protected properties
+
+        /// <summary>
+        /// The first node in the priority list
+        /// </summary>
+        protected LinkedListNode<KeyValuePair<TKey, TValue>> FirstNode
```
Comment (oschwald) [inline MaxMind.Db/CachedDictionary.cs]: I don't think this is used.

File: MaxMind.Db/CachedDictionary.cs [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+        }
+
+        #endregion Public methods
+
+        #region Public properties
+
+        /// <summary>
+        /// Current comparer
+        /// </summary>
+        public IEqualityComparer<TKey> Comparer
```
Comment (oschwald) [inline MaxMind.Db/CachedDictionary.cs]: This and the following three are also not used.

File: MaxMind.Db/Reader.cs [hunk 3]
```diff
@@ -170,8 +212,75 @@ private void Dispose(bool disposing)
...
-            return Find<T>(ipAddress, out prefixLength, injectables);
+            return Find<T>(ipAddress, out _, injectables);
+        }
+
+        /// <summary>
+        /// Get an enumerator that iterates all data nodes in the database
+        /// </summary>
+        /// <param name="cacheSize">The size of the data cache. This can greatly speed enumeration at the cost of memory usage.</param>
+        /// <returns>Enumerator for all data nodes</returns>
+        public IEnumerable<Reader.ReaderIteratorNode<T>> FindAll<T>(int cacheSize = 16384) where T : class
```
Comment (oschwald) [inline MaxMind.Db/Reader.cs]: A test that used something other than `Dictionary<string, object>` might be good. In particular, one using the `TypeHolder` class with the `MaxMind-DB-test-decoder.mmdb` database.

File: MaxMind.Db/CachedDictionary.cs [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+    /// <param name="dictionary">Dictionary</param>
+    /// <param name="key">Key</param>
+    /// <param name="value">Value</param>
+    internal delegate void CachedItemRemovedDelegate<TKey, TValue>(CachedDictionary<TKey, TValue> dictionary, TKey key, TValue value);
+
+    /// <summary>
+    /// A dictionary that caches up to N values in memory. Once the dictionary reaches N count, the last item in the internal list is removed.
+    /// New items are always added to the start of the internal list.
+    /// </summary>
+    internal class CachedDictionary<TKey, TValue> : IDictionary<TKey, TValue>, IDisposable
```
[the remark this answers is not in the document]
Comment (jjxtra) [reply MaxMind.Db/CachedDictionary.cs]: Something I use in other projects, can tag it with MIT license. All code written by me in my spare time for free.

**Comment by jjxtra** [comment]
I've implemented most of these changes. I've moved the code to the benchmark project, but it is commented out for now. Not sure how useful it is at the moment, perhaps someone else can add it later if needed.

File: MaxMind.Db/CachedDictionary.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (oschwald) [reply MaxMind.Db/CachedDictionary.cs]: Thanks! If you are the owner of the code, would you be willing to place it under the Apache 2 license to match the other code in the repo and for the benefit of the patent grant?

[the remark this answers is not in the document]
Comment (jjxtra) [reply MaxMind.Db/CachedDictionary.cs]: Sure no problem

**Comment by oschwald** [comment]
Looks good. Perhaps the benchmark code could be moved to a separate PR or an issue so that we don't have commented out code lying around in  the codebase. Other than that and the license question, I am happy to merge this.

File: MaxMind.Db/CachedDictionary.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jjxtra) [reply MaxMind.Db/CachedDictionary.cs]: It is done.

[the remark this answers is not in the document]
Comment (jjxtra) [reply MaxMind.Db/CachedDictionary.cs]: Commented code eradicated

**Comment by jjxtra** [comment]
Ready to merge when you are.

**Comment by oschwald** [comment]
Looks great! Thanks. I am squashing it so that the binary database does not end up in the Git history.

**Pull request merged**